File: Leetcode/TopInterview150/20.ValidParentheses/ValidParentheses.py
'''
Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Every close bracket has a corresponding open bracket of the same type.
 

Example 1:

Input: s = "()"
Output: true
Example 2:

Input: s = "()[]{}"
Output: true
Example 3:

Input: s = "(]"
Output: false


Constraints:

1 <= s.length <= 104
s consists of parentheses only '()[]{}'.
'''

class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        #Create the stack
        stack = []

        #review the string
        for i in s:
            #for any of these (, {, [
            if i == "(" or i == "{" or i == "[":
                stack.append(i)   #push them in the stack
        
            #if the stack is not empty
            elif len(stack) > 0:
                #review for every closing bracket if the 
                #previous one is the open bracket and pop it
                if i == ")" and stack[-1] == "(":
                    stack.pop()
                elif i == "}" and stack[-1] == "{":
                    stack.pop()
                elif i == "]" and stack[-1] == "[":
                    stack.pop()
                else:
                    return stack.append(i)  #if not just put it in the stack
            else:
                stack.append(i)   #if the stack is empty put first something

        #if the final stack is full a missing bracket exists
        if len(stack) != 0:
            return False

        #if is empty, everything is right
        return True  
            
'''
Time complexity: O(n)
'''

File: Leetcode/TopInterview150/020.ValidParentheses/ValidParentheses.java
/*

Given a string s containing just the characters 
'(', ')', '{', '}', '[' and ']', determine if the
input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Every close bracket has a corresponding open bracket of the
same type.
 

Example 1:

Input: s = "()"

Output: true


Example 2:

Input: s = "()[]{}"

Output: true


Example 3:

Input: s = "(]"

Output: false


Example 4:

Input: s = "([])"

Output: true

 

Constraints:

1 <= s.length <= 104
s consists of parentheses only '()[]{}'.

*/


class Solution {
    public boolean isValid(String s) {

         // Create an arraylist of charaters
        ArrayList<Character> stack = new ArrayList<Character>();

        // iterate through the string
        for (int i = 0; i < s.length(); i++){

            // get each character of the string
            char c = s.charAt(i);

            //If the stack has already something
            if(stack.size() > 0){
                // get last item pushed
                char lp = stack.get(stack.size()-1);
                // if it is ")" or "}" or "]" and last item
                // push it is the opposite remove it
                if (c == ')' && lp == '('){
                    stack.remove(stack.size() - 1);
                }else if(c == '}' && lp == '{'){
                    stack.remove(stack.size() - 1);
                }else if(c == ']' && lp == '['){
                    stack.remove(stack.size() - 1);
                }else{
                    // other case leave it in the stack
                    stack.add(c);
                }
            }else{
                // add something first
                stack.add(c);
            }


        }

        // if we have something that means we have some
        // characters unpaired
        if (stack.size() > 0) return false;

        return true;

    }
}

// Time complexity: O(n)
